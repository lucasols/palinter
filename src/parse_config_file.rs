use serde::{Deserialize, Deserializer};
use std::collections::BTreeMap;

#[derive(Debug, Deserialize, Clone)]
#[serde(untagged)]
pub enum SingleOrMultiple<T> {
    Single(T),
    Multiple(Vec<T>),
}

#[derive(Deserialize, Debug)]
pub struct ParsedFileConditions {
    pub has_extension: Option<SingleOrMultiple<String>>,
    pub has_name: Option<SingleOrMultiple<String>>,
    pub does_not_have_name: Option<SingleOrMultiple<String>>,
}

#[derive(Deserialize, Debug)]
struct FolderConditions {
    has_name: Option<SingleOrMultiple<String>>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct ParsedFileExpect {
    pub name_case_is: Option<String>,
    has_sibling_file: Option<SingleOrMultiple<String>>,
    content_matches_any: Option<SingleOrMultiple<String>>,
    error_msg: Option<String>,
    name_not_matches: Option<SingleOrMultiple<String>>,
    extension_is: Option<SingleOrMultiple<String>>,
}

#[derive(Deserialize, Debug)]
#[serde(untagged)]
pub enum ParsedAnyOr<T> {
    Conditions(T),
    Any(String),
}

#[derive(Deserialize, Debug)]
#[serde(untagged)]
pub enum ParsedRule {
    File {
        #[serde(rename = "if_file")]
        conditions: ParsedAnyOr<ParsedFileConditions>,
        expect: ParsedAnyOr<SingleOrMultiple<ParsedFileExpect>>,
        error_msg: Option<String>,
    },
    Folder {
        #[serde(rename = "if_folder")]
        conditions: ParsedAnyOr<ParsedFileConditions>,
        expect: ParsedAnyOr<SingleOrMultiple<ParsedFileExpect>>,
        error_msg: Option<String>,
    },
    OneOf {
        #[serde(rename = "one_of")]
        rules: Vec<ParsedRule>,
    },
    Block(String),

    #[serde(deserialize_with = "ignore_contents")]
    Error,
}

#[derive(Deserialize, Debug)]
struct FolderConfig {
    has_files: Option<Vec<String>>,
    rules: Option<Vec<ParsedRule>>,

    #[serde(flatten)]
    folders: BTreeMap<String, FolderConfig>,
}

#[derive(Deserialize, Debug)]
pub struct ParsedConfig {
    pub blocks: Option<BTreeMap<String, ParsedRule>>,
    pub global_rules: Option<Vec<ParsedRule>>,
    pub to_have_files: Option<Vec<String>>,

    #[serde(flatten)]
    folders: BTreeMap<String, FolderConfig>,
}

fn ignore_contents<'de, D>(deserializer: D) -> Result<(), D::Error>
where
    D: Deserializer<'de>,
{
    #[allow(unused_must_use)]
    {
        // Ignore any content at this part of the json structure
        deserializer.deserialize_ignored_any(serde::de::IgnoredAny);
    }

    // Return unit as our 'Unknown' variant has no args
    Ok(())
}

pub fn parse_config_string(config: String) -> ParsedConfig {
    serde_json::from_str(&config).expect("Failed to parse config file")
}

pub fn parse_config(config_path: &str) -> ParsedConfig {
    let config = std::fs::read_to_string(config_path).unwrap();

    // TODO: return error instead of unwrap
    let parsed = parse_config_string(config);

    // TODO: validate_parsed_config(&parsed);

    parsed
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_config_works() {
        let config = parse_config("./src/fixtures/config1.json");

        insta::assert_debug_snapshot!(config);
    }
}
