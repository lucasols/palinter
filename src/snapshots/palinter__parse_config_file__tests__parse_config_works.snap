---
source: src/parse_config_file.rs
expression: config
---
ParsedConfig {
    blocks: Some(
        {
            "react_file_name": File {
                conditions: Conditions(
                    ParsedFileConditions {
                        has_extension: Some(
                            Single(
                                "tsx",
                            ),
                        ),
                        has_name: None,
                        does_not_have_name: None,
                    },
                ),
                expect: Conditions(
                    Multiple(
                        [
                            ParsedFileExpect {
                                name_case_is: Some(
                                    "PascalCase",
                                ),
                                has_sibling_file: None,
                                content_matches_any: None,
                                error_msg: Some(
                                    "React files name must be PascalCase",
                                ),
                                name_not_matches: None,
                                extension_is: None,
                            },
                            ParsedFileExpect {
                                name_case_is: None,
                                has_sibling_file: None,
                                content_matches_any: Some(
                                    Multiple(
                                        [
                                            "regex:export const {{fileName}}:? ",
                                        ],
                                    ),
                                ),
                                error_msg: None,
                                name_not_matches: None,
                                extension_is: None,
                            },
                        ],
                    ),
                ),
                error_msg: None,
            },
            "react_sub_file": File {
                conditions: Conditions(
                    ParsedFileConditions {
                        has_extension: None,
                        has_name: Some(
                            Single(
                                "*:PascalCase.(hooks|utils|style|types).ts",
                            ),
                        ),
                        does_not_have_name: None,
                    },
                ),
                expect: Conditions(
                    Single(
                        ParsedFileExpect {
                            name_case_is: None,
                            has_sibling_file: Some(
                                Single(
                                    "{{1}}.tsx",
                                ),
                            ),
                            content_matches_any: None,
                            error_msg: None,
                            name_not_matches: None,
                            extension_is: None,
                        },
                    ),
                ),
                error_msg: None,
            },
        },
    ),
    global_rules: Some(
        [
            File {
                conditions: Conditions(
                    ParsedFileConditions {
                        has_extension: Some(
                            Single(
                                "svg",
                            ),
                        ),
                        has_name: None,
                        does_not_have_name: None,
                    },
                ),
                expect: Conditions(
                    Single(
                        ParsedFileExpect {
                            name_case_is: Some(
                                "kebab-case",
                            ),
                            has_sibling_file: None,
                            content_matches_any: None,
                            error_msg: None,
                            name_not_matches: None,
                            extension_is: None,
                        },
                    ),
                ),
                error_msg: Some(
                    "Svg files should be named in kebab-case",
                ),
            },
        ],
    ),
    to_have_files: None,
    folders: {
        "//src": FolderConfig {
            has_files: None,
            rules: None,
            folders: {
                "//icons": FolderConfig {
                    has_files: None,
                    rules: Some(
                        [
                            File {
                                conditions: Any(
                                    "any",
                                ),
                                expect: Conditions(
                                    Single(
                                        ParsedFileExpect {
                                            name_case_is: None,
                                            has_sibling_file: None,
                                            content_matches_any: None,
                                            error_msg: None,
                                            name_not_matches: None,
                                            extension_is: Some(
                                                Single(
                                                    "svg",
                                                ),
                                            ),
                                        },
                                    ),
                                ),
                                error_msg: Some(
                                    "Only svg files are allowed in /assets/icons",
                                ),
                            },
                            File {
                                conditions: Conditions(
                                    ParsedFileConditions {
                                        has_extension: None,
                                        has_name: None,
                                        does_not_have_name: Some(
                                            Single(
                                                "static-*.ts",
                                            ),
                                        ),
                                    },
                                ),
                                expect: Conditions(
                                    Multiple(
                                        [
                                            ParsedFileExpect {
                                                name_case_is: None,
                                                has_sibling_file: None,
                                                content_matches_any: Some(
                                                    Multiple(
                                                        [
                                                            "currentColor",
                                                        ],
                                                    ),
                                                ),
                                                error_msg: Some(
                                                    "Svg should have at least one color set as `currentColor`, if the icon is static and don't have any dynamic color prefix the name with `static-`",
                                                ),
                                                name_not_matches: None,
                                                extension_is: None,
                                            },
                                            ParsedFileExpect {
                                                name_case_is: None,
                                                has_sibling_file: None,
                                                content_matches_any: Some(
                                                    Multiple(
                                                        [
                                                            "viewBox=\"0 0 24 24\"",
                                                            "viewBox=\"0 0 48 48\"",
                                                        ],
                                                    ),
                                                ),
                                                error_msg: Some(
                                                    "Svg should have a viewBox attribute with 24x24 or 48x48, check if you are exporting the correct svg from figma",
                                                ),
                                                name_not_matches: None,
                                                extension_is: None,
                                            },
                                        ],
                                    ),
                                ),
                                error_msg: None,
                            },
                        ],
                    ),
                    folders: {},
                },
                "//pages": FolderConfig {
                    has_files: None,
                    rules: Some(
                        [
                            OneOf {
                                rules: [
                                    Block(
                                        "react_component",
                                    ),
                                    File {
                                        conditions: Conditions(
                                            ParsedFileConditions {
                                                has_extension: Some(
                                                    Single(
                                                        "tsx",
                                                    ),
                                                ),
                                                has_name: None,
                                                does_not_have_name: None,
                                            },
                                        ),
                                        expect: Conditions(
                                            Multiple(
                                                [
                                                    ParsedFileExpect {
                                                        name_case_is: Some(
                                                            "kebab-case",
                                                        ),
                                                        has_sibling_file: None,
                                                        content_matches_any: None,
                                                        error_msg: Some(
                                                            "React files name must be PascalCase",
                                                        ),
                                                        name_not_matches: None,
                                                        extension_is: None,
                                                    },
                                                    ParsedFileExpect {
                                                        name_case_is: None,
                                                        has_sibling_file: None,
                                                        content_matches_any: Some(
                                                            Multiple(
                                                                [
                                                                    "line_regex:^export const {{fileName:PascalCase}}Page:? ",
                                                                ],
                                                            ),
                                                        ),
                                                        error_msg: None,
                                                        name_not_matches: None,
                                                        extension_is: None,
                                                    },
                                                ],
                                            ),
                                        ),
                                        error_msg: None,
                                    },
                                ],
                            },
                            File {
                                conditions: Conditions(
                                    ParsedFileConditions {
                                        has_extension: None,
                                        has_name: Some(
                                            Single(
                                                "*:kebab-case.(hooks|utils|style|types).ts",
                                            ),
                                        ),
                                        does_not_have_name: None,
                                    },
                                ),
                                expect: Conditions(
                                    Single(
                                        ParsedFileExpect {
                                            name_case_is: None,
                                            has_sibling_file: Some(
                                                Single(
                                                    "$1.tsx",
                                                ),
                                            ),
                                            content_matches_any: None,
                                            error_msg: None,
                                            name_not_matches: None,
                                            extension_is: None,
                                        },
                                    ),
                                ),
                                error_msg: None,
                            },
                            Folder {
                                conditions: Conditions(
                                    ParsedFileConditions {
                                        has_extension: None,
                                        has_name: None,
                                        does_not_have_name: None,
                                    },
                                ),
                                expect: Conditions(
                                    Single(
                                        ParsedFileExpect {
                                            name_case_is: None,
                                            has_sibling_file: None,
                                            content_matches_any: None,
                                            error_msg: None,
                                            name_not_matches: None,
                                            extension_is: None,
                                        },
                                    ),
                                ),
                                error_msg: None,
                            },
                            Folder {
                                conditions: Conditions(
                                    ParsedFileConditions {
                                        has_extension: None,
                                        has_name: None,
                                        does_not_have_name: None,
                                    },
                                ),
                                expect: Conditions(
                                    Single(
                                        ParsedFileExpect {
                                            name_case_is: None,
                                            has_sibling_file: None,
                                            content_matches_any: None,
                                            error_msg: None,
                                            name_not_matches: None,
                                            extension_is: None,
                                        },
                                    ),
                                ),
                                error_msg: None,
                            },
                        ],
                    ),
                    folders: {
                        "//_shell": FolderConfig {
                            has_files: Some(
                                [
                                    "Shell.tsx",
                                ],
                            ),
                            rules: None,
                            folders: {},
                        },
                    },
                },
                "//stores": FolderConfig {
                    has_files: Some(
                        [
                            "collectionStore.example.ts",
                            "documentStore.example.ts",
                            "listQueryStore.example.ts",
                        ],
                    ),
                    rules: Some(
                        [
                            File {
                                conditions: Any(
                                    "any",
                                ),
                                expect: Conditions(
                                    Single(
                                        ParsedFileExpect {
                                            name_case_is: Some(
                                                "camelCase",
                                            ),
                                            has_sibling_file: None,
                                            content_matches_any: None,
                                            error_msg: None,
                                            name_not_matches: None,
                                            extension_is: None,
                                        },
                                    ),
                                ),
                                error_msg: None,
                            },
                            Folder {
                                conditions: Any(
                                    "any",
                                ),
                                expect: Conditions(
                                    Single(
                                        ParsedFileExpect {
                                            name_case_is: Some(
                                                "camelCase",
                                            ),
                                            has_sibling_file: None,
                                            content_matches_any: None,
                                            error_msg: None,
                                            name_not_matches: None,
                                            extension_is: None,
                                        },
                                    ),
                                ),
                                error_msg: None,
                            },
                            File {
                                conditions: Conditions(
                                    ParsedFileConditions {
                                        has_extension: None,
                                        has_name: Some(
                                            Single(
                                                "*Store.utils.ts",
                                            ),
                                        ),
                                        does_not_have_name: None,
                                    },
                                ),
                                expect: Conditions(
                                    Single(
                                        ParsedFileExpect {
                                            name_case_is: None,
                                            has_sibling_file: Some(
                                                Single(
                                                    "$1Store.ts",
                                                ),
                                            ),
                                            content_matches_any: None,
                                            error_msg: None,
                                            name_not_matches: None,
                                            extension_is: None,
                                        },
                                    ),
                                ),
                                error_msg: None,
                            },
                            File {
                                conditions: Conditions(
                                    ParsedFileConditions {
                                        has_extension: None,
                                        has_name: Some(
                                            Single(
                                                "*(Doc|List|Query).(utils|actions).ts",
                                            ),
                                        ),
                                        does_not_have_name: None,
                                    },
                                ),
                                expect: Conditions(
                                    Single(
                                        ParsedFileExpect {
                                            name_case_is: None,
                                            has_sibling_file: Some(
                                                Single(
                                                    "$1$2.ts",
                                                ),
                                            ),
                                            content_matches_any: None,
                                            error_msg: None,
                                            name_not_matches: None,
                                            extension_is: None,
                                        },
                                    ),
                                ),
                                error_msg: None,
                            },
                            File {
                                conditions: Conditions(
                                    ParsedFileConditions {
                                        has_extension: None,
                                        has_name: Some(
                                            Single(
                                                "*Doc.ts",
                                            ),
                                        ),
                                        does_not_have_name: None,
                                    },
                                ),
                                expect: Conditions(
                                    Single(
                                        ParsedFileExpect {
                                            name_case_is: None,
                                            has_sibling_file: None,
                                            content_matches_any: Some(
                                                Multiple(
                                                    [
                                                        "export const $1Doc = createDocumentStore<",
                                                    ],
                                                ),
                                            ),
                                            error_msg: None,
                                            name_not_matches: None,
                                            extension_is: None,
                                        },
                                    ),
                                ),
                                error_msg: None,
                            },
                            File {
                                conditions: Conditions(
                                    ParsedFileConditions {
                                        has_extension: None,
                                        has_name: Some(
                                            Single(
                                                "*List.ts",
                                            ),
                                        ),
                                        does_not_have_name: None,
                                    },
                                ),
                                expect: Conditions(
                                    Single(
                                        ParsedFileExpect {
                                            name_case_is: None,
                                            has_sibling_file: None,
                                            content_matches_any: Some(
                                                Multiple(
                                                    [
                                                        "export const {{1}}List = createListQueryStore<",
                                                    ],
                                                ),
                                            ),
                                            error_msg: None,
                                            name_not_matches: None,
                                            extension_is: None,
                                        },
                                    ),
                                ),
                                error_msg: None,
                            },
                            File {
                                conditions: Conditions(
                                    ParsedFileConditions {
                                        has_extension: None,
                                        has_name: Some(
                                            Single(
                                                "*Doc.ts",
                                            ),
                                        ),
                                        does_not_have_name: None,
                                    },
                                ),
                                expect: Conditions(
                                    Single(
                                        ParsedFileExpect {
                                            name_case_is: None,
                                            has_sibling_file: None,
                                            content_matches_any: Some(
                                                Multiple(
                                                    [
                                                        "export const $1Doc = createDocumentStore<",
                                                    ],
                                                ),
                                            ),
                                            error_msg: None,
                                            name_not_matches: None,
                                            extension_is: None,
                                        },
                                    ),
                                ),
                                error_msg: None,
                            },
                            Folder {
                                conditions: Conditions(
                                    ParsedFileConditions {
                                        has_extension: None,
                                        has_name: None,
                                        does_not_have_name: None,
                                    },
                                ),
                                expect: Conditions(
                                    Single(
                                        ParsedFileExpect {
                                            name_case_is: None,
                                            has_sibling_file: None,
                                            content_matches_any: None,
                                            error_msg: None,
                                            name_not_matches: None,
                                            extension_is: None,
                                        },
                                    ),
                                ),
                                error_msg: None,
                            },
                        ],
                    ),
                    folders: {},
                },
            },
        },
    },
}
